python/tvm/contrib/relay_viz/__init__.py

```diff
@@ -0,0 +1,160 @@
+                graph_names.insert(0, gv_name.name_hint)
+            else:
+                graph_names.append(gv_name.name_hint)
+
+        for name in graph_names:
+            # clear previous graph
+            self._node_to_id = {}
+            relay.analysis.post_order_visit(
+                relay_mod[name],
+                lambda node: self._traverse_expr(node),  # pylint: disable=unnecessary-lambda
+            )
+            graph = self._plotter.create_graph(name)
+            # shallow copy to prevent callback modify self._node_to_id
+            self._render_cb(graph, copy.copy(self._node_to_id), self._relay_param)
+
+    def _traverse_expr(self, node):
+        # based on https://github.com/apache/tvm/pull/4370
+        if node in self._node_to_id:
+            return
+        self._node_to_id[node] = len(self._node_to_id)
+
+    def _render_cb(self, graph, node_to_id, relay_param):
+        """a callback to Add nodes and edges to the graph.
+
+        Parameters
+        ----------
+        graph : class plotter.Graph
+
+        node_to_id : Dict[relay.expr, int]
+
+        relay_param : Dict[string, NDarray]
+        """
+        # Based on https://tvm.apache.org/2020/07/14/bert-pytorch-tvm
+        unknown_type = "unknown"
+        for node, node_id in node_to_id.items():
+            if type(node) in self._render_rules:  # pylint: disable=unidiomatic-typecheck
+                graph_info, edge_info = self._render_rules[type(node)](
+                    node, relay_param, node_to_id
+                )
+                if graph_info:
+                    graph.node(*graph_info)
+                for edge in edge_info:
+                    graph.edge(*edge)
+            else:
+                unknown_info = "Unknown node: {}".format(type(node))
+                _LOGGER.warning(unknown_info)
+                graph.node(node_id, unknown_type, unknown_info)
+
+    def render(self, filename):
+        return self._plotter.render(filename=filename)
+
+
+def get_plotter_and_render_rules(backend):
+    """Specify the Plottor and its render rules
+
+    Parameters
+        ----------
+        backend: PlotterBackend or a tuple
+                        PlotterBackend: The backend of plotting. Default "bokeh"
+                        Tuple: A tuple with two arguments. First is user-defined Plotter, \
+                               the second is user-defined RenderCallback
+    """
+    if type(backend) is tuple and len(backend) == 2:  # pylint: disable=unidiomatic-typecheck
+        if not isinstance(backend[0], Plotter):
+            raise ValueError("First elemnet of the backend should be a plotter")
+        plotter = backend[0]
+        if not isinstance(backend[1], RenderCallback):
+            raise ValueError("Second elemnet of the backend should be a callback")
+        render = backend[1]
+        render_rules = render.get_rules()
+        return plotter, render_rules
+
+    if backend in PlotterBackend:
```

Comment:
rather than handle the good case here, handle the bad case and bail:

```
if backend not in PlotterBackend:
  raise ValueError(...)
```

then the rest of the function can un-indent.